### ISSUE TITLE
SIGABRT in msgDestruct (rsyslog-8.30)

### ISSUE THREAD

**Issue by hrak** [rsyslog/rsyslog#1997]
Seeing these on a fairly frequent basis with rsyslog 8.30 on Ubuntu 14.04 (package 8.30.0-0adiscon2trusty1)

```
*** Error in `/usr/sbin/rsyslogd': double free or corruption (out): 0x00007fffd401fad0 ***

Program received signal SIGABRT, Aborted.
[Switching to Thread 0x7fffd3bde700 (LWP 24159)]
0x00007ffff6bf0c37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56	../nptl/sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  0x00007ffff6bf0c37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff6bf4028 in __GI_abort () at abort.c:89
#2  0x00007ffff6c2d2a4 in __libc_message (do_abort=do_abort@entry=1, fmt=fmt@entry=0x7ffff6d3f310 "*** Error in `%s': %s: 0x%s ***\n")
    at ../sysdeps/posix/libc_fatal.c:175
#3  0x00007ffff6c3982e in malloc_printerr (ptr=<optimized out>, str=0x7ffff6d3f440 "double free or corruption (out)", action=1) at malloc.c:4998
#4  _int_free (av=<optimized out>, p=<optimized out>, have_lock=0) at malloc.c:3842
#5  0x00000000004316bc in msgDestruct (ppThis=ppThis@entry=0x7fffd3bddb78) at msg.c:970
#6  0x000000000044de72 in DeleteProcessedBatch (pThis=pThis@entry=0x6d2ed0, pBatch=pBatch@entry=0x6f5430) at queue.c:1648
#7  0x000000000044e45f in DequeueConsumableElements (pSkippedMsgs=<optimized out>, piRemainingQueueSize=<synthetic pointer>, pWti=<optimized out>,
    pThis=0x6d2ed0) at queue.c:1685
#8  DequeueConsumable (pThis=pThis@entry=0x6d2ed0, pWti=pWti@entry=0x6f5400, pSkippedMsgs=pSkippedMsgs@entry=0x7fffd3bddc6c) at queue.c:1774
#9  0x000000000044e802 in DequeueForConsumer (pSkippedMsgs=0x7fffd3bddc6c, pWti=0x6f5400, pThis=0x6d2ed0) at queue.c:1919
#10 ConsumerReg (pThis=0x6d2ed0, pWti=0x6f5400) at queue.c:1975
#11 0x000000000044b3ce in wtiWorker (pThis=pThis@entry=0x6f5400) at wti.c:365
#12 0x000000000044a425 in wtpWorker (arg=0x6f5400) at wtp.c:417
#13 0x00007ffff79ab184 in start_thread (arg=0x7fffd3bde700) at pthread_create.c:312
#14 0x00007ffff6cb7ffd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

I can provide more info if needed

**Comment by rgerhards** [comment]
Could you re-try against git master branch?

**Comment by rgerhards** [comment]
flagged just as "potential" bug because it may already be gone. The thing for sure is a bug ;-)

**Comment by hrak** [comment]
i have upgraded one of our prod log forwarders to 8.31, will report back tomorrow with results

**Comment by hrak** [comment]
I'm still seeing pretty frequent crashes of rsyslog on several fronts. Since the upgrade yesterday this is the list:

```
[Tue Dec  5 15:53:08 2017] init: rsyslog main process (27403) killed by ABRT signal
[Tue Dec  5 16:22:49 2017] init: rsyslog main process (23359) killed by ABRT signal
[Tue Dec  5 16:43:42 2017] init: rsyslog main process (25336) killed by ABRT signal
[Tue Dec  5 17:19:37 2017] init: rsyslog main process (25890) killed by ABRT signal
[Tue Dec  5 17:31:05 2017] init: rsyslog main process (28020) killed by ABRT signal
[Tue Dec  5 17:42:58 2017] init: rsyslog main process (29568) killed by ABRT signal
[Tue Dec  5 17:49:17 2017] init: rsyslog main process (29869) killed by SEGV signal
[Tue Dec  5 18:00:06 2017] init: rsyslog main process (30043) killed by SEGV signal
[Tue Dec  5 18:07:37 2017] init: rsyslog main process (31576) killed by ABRT signal
[Tue Dec  5 18:10:03 2017] init: rsyslog main process (31787) killed by ABRT signal
[Tue Dec  5 18:42:14 2017] init: rsyslog main process (31859) killed by SEGV signal
[Tue Dec  5 19:37:11 2017] init: rsyslog main process (1480) killed by ABRT signal
[Tue Dec  5 19:43:26 2017] init: rsyslog main process (4065) killed by ABRT signal
[Tue Dec  5 19:47:55 2017] init: rsyslog main process (5487) killed by SEGV signal
[Tue Dec  5 19:59:17 2017] init: rsyslog main process (5624) killed by ABRT signal
[Tue Dec  5 20:10:38 2017] init: rsyslog main process (5922) killed by ABRT signal
[Tue Dec  5 21:42:43 2017] init: rsyslog main process (6204) killed by ABRT signal
[Tue Dec  5 22:00:48 2017] init: rsyslog main process (12239) killed by ABRT signal
[Tue Dec  5 22:09:59 2017] init: rsyslog main process (13961) killed by ABRT signal
[Tue Dec  5 23:55:11 2017] init: rsyslog main process (14211) killed by ABRT signal
[Wed Dec  6 00:13:40 2017] init: rsyslog main process (20452) killed by SEGV signal
[Wed Dec  6 00:30:29 2017] init: rsyslog main process (22156) killed by ABRT signal
[Wed Dec  6 01:00:49 2017] init: rsyslog main process (22603) killed by SEGV signal
[Wed Dec  6 01:04:05 2017] init: rsyslog main process (24557) killed by ABRT signal
[Wed Dec  6 01:20:04 2017] init: rsyslog main process (24680) killed by SEGV signal
[Wed Dec  6 01:38:54 2017] init: rsyslog main process (26331) killed by ABRT signal
[Wed Dec  6 01:45:29 2017] init: rsyslog main process (28068) killed by ABRT signal
[Wed Dec  6 01:47:43 2017] init: rsyslog main process (28248) killed by ABRT signal
[Wed Dec  6 01:57:18 2017] init: rsyslog main process (28326) killed by ABRT signal
[Wed Dec  6 02:24:50 2017] init: rsyslog main process (28586) killed by SEGV signal
[Wed Dec  6 02:29:20 2017] init: rsyslog main process (30552) killed by ABRT signal
[Wed Dec  6 02:34:12 2017] init: rsyslog main process (30685) killed by ABRT signal
[Wed Dec  6 02:55:28 2017] init: rsyslog main process (30830) killed by ABRT signal
[Wed Dec  6 03:14:17 2017] init: rsyslog main process (32631) killed by SEGV signal
[Wed Dec  6 03:32:23 2017] init: rsyslog main process (1928) killed by ABRT signal
[Wed Dec  6 03:41:00 2017] init: rsyslog main process (2379) killed by ABRT signal
[Wed Dec  6 03:45:50 2017] init: rsyslog main process (2627) killed by ABRT signal
[Wed Dec  6 04:06:51 2017] init: rsyslog main process (4012) killed by SEGV signal
[Wed Dec  6 05:08:15 2017] init: rsyslog main process (4530) killed by ABRT signal
[Wed Dec  6 05:09:41 2017] init: rsyslog main process (8574) killed by SEGV signal
[Wed Dec  6 05:18:33 2017] init: rsyslog main process (8686) killed by ABRT signal
[Wed Dec  6 05:58:31 2017] init: rsyslog main process (8982) killed by ABRT signal
[Wed Dec  6 05:59:16 2017] init: rsyslog main process (12461) killed by ABRT signal
[Wed Dec  6 06:30:49 2017] init: rsyslog main process (12505) killed by ABRT signal
[Wed Dec  6 06:36:11 2017] init: rsyslog main process (14641) killed by ABRT signal
[Wed Dec  6 06:39:16 2017] init: rsyslog main process (14822) killed by ABRT signal
[Wed Dec  6 06:54:42 2017] init: rsyslog main process (14925) killed by SEGV signal
[Wed Dec  6 07:18:12 2017] init: rsyslog main process (15317) killed by SEGV signal
[Wed Dec  6 07:22:21 2017] init: rsyslog main process (17149) killed by SEGV signal
[Wed Dec  6 07:30:44 2017] init: rsyslog main process (17274) killed by ABRT signal
[Wed Dec  6 07:30:54 2017] init: rsyslog main process (18739) killed by SEGV signal
[Wed Dec  6 07:34:12 2017] init: rsyslog main process (18778) killed by SEGV signal
[Wed Dec  6 07:34:40 2017] init: rsyslog main process (18900) killed by ABRT signal
[Wed Dec  6 07:59:19 2017] init: rsyslog main process (18939) killed by SEGV signal
```

Also (unrelated) seeing segfaults in imudp (error 4 and error 6) in both 8.30 and 8.31:

```
[Fri Dec  1 13:20:20 2017] in:imudp[2283]: segfault at 10 ip 000000000043a136 sp 00007fca6f7f9c40 error 4 in rsyslogd[400000+91000]
[Fri Dec  1 18:35:02 2017] in:imudp[20918]: segfault at 40 ip 000000000043169c sp 00007f23569dcc70 error 6 in rsyslogd[400000+91000]
[Fri Dec  1 23:15:03 2017] in:imudp[5623]: segfault at 40 ip 000000000043169c sp 00007fdad8870c70 error 6 in rsyslogd[400000+91000]
[Sat Dec  2 15:58:59 2017] in:imudp[7707]: segfault at 10 ip 000000000043a136 sp 00007fe70ebf7c40 error 4 in rsyslogd[400000+91000]
[Sun Dec  3 03:30:25 2017] in:imudp[18500]: segfault at 40 ip 000000000043169c sp 00007f50633f8c70 error 6 in rsyslogd[400000+91000]
[Sun Dec  3 07:44:49 2017] in:imudp[1424]: segfault at 40 ip 000000000043169c sp 00007fb64fbfadb0 error 6 in rsyslogd[400000+91000]
[Sun Dec  3 17:25:53 2017] in:imudp[8168]: segfault at 10 ip 000000000043a136 sp 00007f65c2ff7d80 error 4 in rsyslogd[400000+91000]
[Mon Dec  4 12:55:40 2017] in:imudp[17412]: segfault at 40 ip 000000000043169c sp 00007fe58e9dbdb0 error 6 in rsyslogd[400000+91000]
[Tue Dec  5 13:22:24 2017] in:imudp[14343]: segfault at 40 ip 000000000043169c sp 00007f508d458c70 error 6 in rsyslogd[400000+91000]
[Tue Dec  5 17:49:17 2017] in:imudp[29875]: segfault at 40 ip 000000000043134c sp 00007f58b128dc70 error 6 in rsyslogd[400000+90000]
```

I will run it in gdb & valgrind again to see if i can gather some more info.

**Comment by rgerhards** [comment]
Do you have any valgrind results?

**Comment by hrak** [comment]
Here are the results (rsyslog 8.31.0-0adiscon3trusty1 w/ libfastjson4 0.99.8-adiscon1trusty1)

Also caught another (unrelated?) segfault:

```
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff182a700 (LWP 6030)]
__memcpy_sse2_unaligned () at ../sysdeps/x86_64/multiarch/memcpy-sse2-unaligned.S:152
152	../sysdeps/x86_64/multiarch/memcpy-sse2-unaligned.S: No such file or directory.
(gdb) bt
#0  __memcpy_sse2_unaligned () at ../sysdeps/x86_64/multiarch/memcpy-sse2-unaligned.S:152
#1  0x00000000004354b4 in memcpy (__len=<optimized out>, __src=<optimized out>, __dest=0x7fffd8016290)
    at /usr/include/x86_64-linux-gnu/bits/string3.h:51
#2  MsgReplaceMSG (pThis=pThis@entry=0x7fffd8016020,
    pszMSG=pszMSG@entry=0x7ffff18269a0 " message repeated 2 times: [ : IP  InetServer:s=10.22.215.187 d=10.32.255.6(vl331) g=10.22.215.190 xmit ]", lenMSG=105) at msg.c:2856
#3  0x00000000004537e0 in ratelimitGenRepMsg (ratelimit=<optimized out>, ratelimit=<optimized out>) at ratelimit.c:66
#4  0x0000000000453b0c in doLastMessageRepeatedNTimes (ppRepMsg=0x7ffff1827428, pMsg=0x7fffd80163e0, ratelimit=0x6d53d0) at ratelimit.c:99
#5  ratelimitMsg (ratelimit=0x6d53d0, pMsg=pMsg@entry=0x7fffd80163e0, ppRepMsg=ppRepMsg@entry=0x7ffff1827428) at ratelimit.c:236
#6  0x0000000000453ce4 in ratelimitAddMsg (ratelimit=<optimized out>, pMultiSub=pMultiSub@entry=0x7ffff18274f0, pMsg=0x7fffd80163e0)
    at ratelimit.c:273
#7  0x00007ffff5b667d5 in processPacket (socklen=<optimized out>, multiSub=0x7ffff18274f0, frominet=<optimized out>,
    ttGenTime=<optimized out>, stTime=0x7ffff1827500, lenRcvBuf=<optimized out>,
    rcvBuf=0x7ffff7fe54ee "<15> Dec 20 08:50:31 10.22.215.187 IP : IP  InetServer:s=10.22.215.187 d=10.32.255.6(vl331) g=10.22.215.190 xmit 94.d5.47.a8.6from LEARNING to FORWARDINGGG", pbIsPermitted=<synthetic pointer>, frominetPrev=0x7ffff18295a0, lstn=0x6d2a90)
    at imudp.c:455
#8  processSocket (pbIsPermitted=<synthetic pointer>, frominetPrev=0x7ffff18295a0, lstn=0x6d2a90, pWrkr=0x7ffff5d698c0 <wrkrInfo>)
    at imudp.c:538
#9  rcvMainLoop (pWrkr=pWrkr@entry=0x7ffff5d698c0 <wrkrInfo>) at imudp.c:824
#10 0x00007ffff5b66be0 in wrkr (myself=0x7ffff5d698c0 <wrkrInfo>) at imudp.c:1221
#11 0x00007ffff79ab184 in start_thread (arg=0x7ffff182a700) at pthread_create.c:312
#12 0x00007ffff6cb7ffd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

I stripped a bunch of unrelated libssl/libcrypto stuff from the valgrind output.

[valgrind.txt](https://github.com/rsyslog/rsyslog/files/1574921/valgrind.txt)

**Comment by rgerhards** [comment]
thx

**Comment by rgerhards** [comment]
The unrelated segfault was this one: https://github.com/rsyslog/rsyslog/pull/2183

@hrak I have seen that I unfortunately do not have your configuration on the ticket. Right now, this problem may either be

* configuration induced
* caused by libfastjson < 0.99.8

If you could add your config, that would be great. Also it might be a good idea to install libfastjson 0.99.7, which has two important bug fixes. It is questionable if they affect your problem (as I don't have the config) but if you use variables and especially mmjsonparse, it might be the case.

**Comment by rgerhards** [comment]
@hrak Well... looking in more depth at the valgrind log, it looks like the issue probably is solved by a combination of #2183 and libfastjson 0.99.8. I definitely see some reports that point to the presence of #2183. That did not always lead to immediate segfault, but also to memory corruption which later could cause segfault at unrelated code regions.

Plan forward: please upgrade to
* libfastjson 0.99.8
* git master branch rsyslog (via build from source or Ubuntu daily package)

for now, I close the issue. If the problem persists, let me know and I'll swiftly re-open.
